src/coreclr/jit/promotion.cpp

```diff
@@ -378,52 +385,78 @@ class LocalUses
                 return false;
             }
 
-            countOverlappedCallsWtd += otherAccess.CountCallArgsWtd;
+            countOverlappedCallArgWtd += otherAccess.CountCallArgsWtd;
             countOverlappedReturnsWtd += otherAccess.CountReturnsWtd;
             countOverlappedRetbufsWtd += otherAccess.CountPassedAsRetbufWtd;
-            countOverlappedAssignmentDestinationWtd += otherAccess.CountAssignmentDestinationWtd;
-            countOverlappedAssignmentSourceWtd += otherAccess.CountAssignmentSourceWtd;
+            countOverlappedAssignedFromCallWtd += otherAccess.CountAssignedFromCallWtd;
+            countOverlappedDecomposableAssignmentWtd +=
+                (otherAccess.CountAssignmentDestinationWtd + otherAccess.CountAssignmentSourceWtd -
+                 otherAccess.CountAssignedFromCallWtd);
         }
 
-        // TODO-CQ: Tune the following heuristics. Currently they are based on
-        // x64 code size although using BB weights when available. This mixing
-        // does not make sense.
         weight_t costWithout = 0;
 
-        // A normal access without promotion looks like:
-        // mov reg, [reg+offs]
-        // It may also be contained. Overall we are going to cost each use of
-        // an unpromoted local at 6.5 bytes.
-        // TODO-CQ: We can make much better guesses on what will and won't be contained.
-        costWithout += access.CountWtd * 6.5;
+        // We cost any normal access (which is a struct load or store) without promotion at 3 cycles.
+        costWithout += access.CountWtd * 3;
```

Comment:
Consider making all these weight factors symbolic so later on we can vary them more readily.

Reply:
Makes sense, I was initially using the perfscore constants here but noticed some oddities (e.g. stack writes are cheaper on arm64 than stack reads but they are the same cost for xarch). I can add some names the next time around.